### PR TITLE
Feature/x509 san

### PR DESCRIPTION
**Pull request by JesusMcCloud**
Tests are missing, kmp-crypto is duct-tape-patched, so I am making it a draft.

@nodh Let me know what else you require before we can merge, given we want to get it deployed ASAP

### CODE REVIEW

File: vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt [hunk 1]
```diff
@@ -252,98 +252,209 @@ class OidcSiopWallet(
...
+        if (params.parameters.clientIdScheme == OpenIdConstants.ClientIdSchemes.REDIRECT_URI
+            && (params.parameters.clientMetadata == null && params.parameters.clientMetadataUri == null)
         ) {
             return KmmResult.failure<AuthenticationResponseParameters>(OAuth2Exception(Errors.INVALID_REQUEST))
                 .also { Napier.w("client_id_scheme is redirect_uri, but metadata is not set") }
         }
+
+        //TODO: was the JWsSigned already cryptographically verified? esp. th x5c?
+        if ((params.parameters.clientIdScheme == OpenIdConstants.ClientIdSchemes.X509_SAN_DNS)
+            || (params.parameters.clientIdScheme == OpenIdConstants.ClientIdSchemes.X509_SAN_DNS)
```
Comment (nodh) [inline vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt:267]: aren't both checks the same?

File: vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt [hunk 2]
```diff
@@ -252,98 +252,209 @@ class OidcSiopWallet(
...
+                        "client_id_scheme is ${
+                            params.parameters.clientIdScheme
+                        }, but metadata is not set and no x5c certificate chain is present in the original authn request"
+                    )
+                }
+            else { //basic checks done
+                val leaf = params.source.header.certificateChain!!.leaf
+                if (leaf.tbsCertificate.extensions == null || leaf.tbsCertificate.extensions!!.isEmpty()) {
+                    return KmmResult.failure<AuthenticationResponseParameters>(OAuth2Exception(Errors.INVALID_REQUEST))
+                        .also {
```
Comment (nodh) [inline vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt:285]: please reformat the logging statements to be ... somehow on one lin, so we can skim through the code more easily

**Review by nodh — CHANGES_REQUESTED** [review]
Please rebase your branch onto `develop`.
The implementation in `OdicSiopWallet` looks okay, but we'll also need an option to specify the correct metadata in `OidcSiopVerifier`

**Comment by JesusMcCloud** [comment]
Amended the verifier and (well the best way to put it is really:) changed the code until all the tests went green.

image

Still, I think it is a solid foundation to go through the diff and compare it to the spec.

**Review by nodh — APPROVED** [review]

**Pull request merged**